tests/test-dladdr.c

```diff
@@ -0,0 +1,195 @@
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#elif defined(_MSC_VER)
+/* disable warning C4273 inconsistent dll linkage */
+#pragma warning(push)
+#pragma warning(disable: 4273)
+#endif
+/* fix for i586-mingw32msvc-gcc */
+#ifndef WINBASEAPI
+#define WINBASEAPI __declspec(dllimport)
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+WINBASEAPI HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);
+WINBASEAPI int __cdecl atoi(const char *_Str);
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+#endif
+
+#ifdef _WIN32
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#define PassWithDifferentAddressOnWin PassWithDifferentAddress
+#else
+#define PassWithoutSymbolOnWin Pass
+#define PassWithDifferentAddressOnWin Pass
+#endif
+
+#ifdef DLFCN_WIN32_SHARED
+#define PassWithSymbolOnSharedBuild Pass
+#else
+#define PassWithSymbolOnSharedBuild PassWithoutSymbol
+#endif
+
+#define UNUSED(x) (void)(x)
+
+int main(int argc, char **argv)
+{
+    UNUSED(argv);
+
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    /* null pointer */
+    result = check_dladdr((void*)0, "null" , NoInfo);
+    /* invalid pointer */
+    result |= check_dladdr((void*)0x125, "invalid pointer" , NoInfo);
+    /* -ldl */
+    result |= check_dladdr( ((void*)dladdr), "dladdr" , PassWithSymbolOnSharedBuild );
+    /* -ldl */
+    result |= check_dladdr( (void*)dlopen, "dlopen", PassWithSymbolOnSharedBuild );
+    /* -lglibc */
+    result |= check_dladdr( (void*)atoi, "atoi", Pass );
+    /* test-dladdr */
+    result |= check_dladdr( (void*)main, "main", PassWithoutSymbolOnWin );
+    /* offsets */
+    result |= check_dladdr( (char*)dladdr-6, "dladdr-6", Fail );
```

Comment:
I was thinking more about this test and I'm not sure what this test should check. Because it checks address before dladdr() function. So for sure it is not content of dladdr() function (as it is for +6) but we do not know what is before dladdr(). There may be another function and in this case it should return symbol name of that "another function". Or there can be no function or even no allocated memory and in this case this test should fail. So maybe the best would be to remove this `-6` test as it is checking memory about which we do not know anything.

For address `dladdr+6` we for sure know that it is inside dladdr() function.